### ISSUE TITLE
Cursor should be a crosshairs cursor for points 'add' mode

### ISSUE THREAD

**Issue by jni** [napari/napari#2142]
## 🚀 Feature

It's right there in the icon for "add"! 😂 

Less flippantly, when adding points, the precision of the "hand" icon is very much lacking, see https://github.com/jni/affinder/issues/14 for example. A crosshairs cursor would make it much clearer where exactly you are placing a point.

**Comment by sofroniewn** [comment]
We have a `cross` i think it is called as an option, so should be easy to change!

**Comment by sofroniewn** [comment]
Changing line here https://github.com/napari/napari/blob/c5d10bc6f1806d82fead9a94e36cf9cf37f34a81/napari/layers/points/points.py#L1359 to `self.cursor = 'cross'` should do it I think

**Comment by VolkerH** [comment]
Ok, I tried setting the cursor to 'cross' using the ipython console.
This indeed changes to some crosshairs, but not in a way that offers an improvement for accurate selection.
I don't know to what degree this is system dependent as Qt uses some native gui stuff for certain things, but on my Ubuntu the `crosshairs` is a solid (not transparent) black cross that is a few pixels wide with a white boundary. (My linux screenshot tool doesn't offer the option to include the cursor, otherwise I would show it here)

I envision something more akin to this:

![image](https://user-images.githubusercontent.com/2210044/105470027-5756f380-5c99-11eb-857b-6896a94432f0.png)

With the particular crosshairs enabled by setting `layer.cursor=cross` it actually becomes harder to see the exact pixel the cursor is pointing to than with the pointing hand.

**Comment by sofroniewn** [comment]
Hmm, we could create our own cross hair in the same way that we create our own circle and square brushes - you'd create something like this https://github.com/napari/napari/blob/afc71799cdb49cca885106dbe60f2a8f9d7f25cf/napari/_qt/utils.py#L75-L87 I think, wouldn't be too hard to programmatically do

**Comment by VolkerH** [comment]
Had a bit of a play. Still needs polishing (doesn't respect canvas boundaries no check for gap size etc).


```python
@lru_cache(maxsize=64)
def crosshair_pixmap(size, gap=2):
    """Create a white/black hollow square pixmap. For use as labels cursor."""
    size = max(int(size), 1)
    pixmap = QPixmap(QSize(size, size))
    pixmap.fill(Qt.transparent)
    painter = QPainter(pixmap)
    painter.setPen(Qt.white)
    painter.drawLine(QPointF(0, size/2), QPointF(size/2-gap,size/2))
    painter.drawLine(QPointF(size/2+gap, size/2), QPointF(size, size/2))
    painter.drawLine(QPointF(size/2, 0), QPointF(size/2,size/2-gap))
    painter.drawLine(QPointF(size/2, size/2+gap), QPointF(size/2, size))
    painter.end()
    return pixmap
```

with a size of 64.

![image](https://user-images.githubusercontent.com/2210044/105546592-dc73f400-5cfd-11eb-8ebf-1543ad23a7e7.png)

**Comment by emilmelnikov** [comment]
I would like to add that a similar problem exists in the brush mode: the cursor is quite hard to see if the brush size is small and the image is zoomed out.

This motivated me to suggest an additional feature for the viewer: press a modifier key to activate a "loupe" mode, in which area around the cursor is zoomed to 100%, while still being able to perform an operation with the selected tool. This is somewhat similar to color pickers in image editors.

![image](https://user-images.githubusercontent.com/1649961/108173086-f7b80080-70fd-11eb-9315-505db173731b.png)

**Comment by sofroniewn** [comment]
Love this suggestion @emilmelnikov, we don't have support for anything else like this right now, but that would be a super nice feature. we can definitely think about how we could support it

**Comment by isabela-pf** [comment]
I found this issue via #3287, but it looks like you've already agreed on what the crosshair styling should look like. Are y'all still interested in seeing other crosshair designs?

**Comment by goanpeca** [comment]
>Are y'all still interested in seeing other crosshair designs?

Always :)

**Comment by isabela-pf** [comment]
I’m back with some simple suggestions. I tried out cursors in other image-focused applications, and based on personal experience precision selection is a pretty common problem.

If it counts for anything, I totally agree with @ above that the loupe approach is the experience that feels the most precise. I also imagine this would take significantly more work to make happen, so I’d think this is a long-term fix.

Since switching out a cursor visual seems easier, here are some thoughts I have on what might improve this cursor experience.

I also agree that the cross with a gap in the middle is the most familiar and easiest to read. The strongest options I have ended up being similar to the shapes discussed above, so the main difference is changing the proportions of cross and box.

Many cursors have a black cursor shape with a thinner white outline to make sure the cursor is visible on dark or light backgrounds. This is more versatile and easier to spot than semi-transparent cursors.

Option 0-12

This is closest to the option above and probably my preference. I think it is the easiest to see on the canvas and makes the selection area clear.

<img width="1088" alt="Empty cross (0-12) Light" src="https://user-images.githubusercontent.com/50221806/136085126-5e8a5a09-d602-4c49-92bf-ad677dbee619.png">

<img width="100" alt="Empty cross (0-12) Dark" src="https://user-images.githubusercontent.com/50221806/136085136-5a38b104-1696-462f-9d39-256f6af10e1a.png">

Option 0-5 

I also wanted to see how necessary the additional box was. This is just a cross with a missing center. It works, but I do find it harder to spot the selection here.

<img width="1088" alt="Empty cross (0-5) Light" src="https://user-images.githubusercontent.com/50221806/136085163-df25750b-d096-4697-a121-28ebd2453131.png">

<img width="100" alt="Empty cross (0-5) Dark" src="https://user-images.githubusercontent.com/50221806/136085174-8d898dfe-0ca5-47bb-8752-715acea797e5.png">

Option 0-16

I don’t think this is as precise a selection as the others, but it errs on the side of easiest to spot. For that reason, I wanted to include it in case it inspired any helpful feedback.

<img width="1088" alt="Empty cross (0-16) Light" src="https://user-images.githubusercontent.com/50221806/136085230-72cda92e-3bd3-4950-b8c4-4e74176cbb8b.png">

<img width="100" alt="Empty cross (0-16) Dark" src="https://user-images.githubusercontent.com/50221806/136085239-13a00203-74b3-4353-bd4f-179568570f8e.png">

Questions I still have:
- Is the cursor shape static or can it/will it vary depending on how big a pixel is displayed for that image. I think I’d want it to be static, but that also seems like it could cause problems in the precision of selection (if an image is higher resolution and doesn’t have visible pixels when opened, for example, then the cursor may not feel any more precise than it currently does).
- What would be a good format for passing on this information and/or what specifications would be relevant for anyone interested in editing the cursor? I’ve never designed for cursors before, so I don’t know.

**Comment by sofroniewn** [comment]
> Is the cursor shape static or can it/will it vary depending on how big a pixel is displayed for that image. I think I’d want it to be static, but that also seems like it could cause problems in the precision of selection (if an image is higher resolution and doesn’t have visible pixels when opened, for example, then the cursor may not feel any more precise than it currently does).

We should be able to swap out the cursor depending on zoom level. I probably wouldn't have too many "levels", but we could have a high mag and a low mag shape.

> What would be a good format for passing on this information and/or what specifications would be relevant for anyone interested in editing the cursor? I’ve never designed for cursors before, so I don’t know.

The best way for us to code up the cursors is what is done here https://github.com/napari/napari/blob/5f1525c394606bfc5193a837f55f1c38cdcbeac7/napari/_qt/utils.py#L151-L178 with the square and circle pixmap. We just need to be able to define every "line" in the cursor as it gets drawn using the Qt primitive drawing operations like rectangle, ellipse etc

**Comment by sofroniewn** [comment]
I like something like the first option best too. It would basically be to take the square pixmap and add the lines coming out from the centers. That shouldn't be too hard to code up as a demo. We could then see how it felt

**Comment by isabela-pf** [comment]
I agree I wouldn't want too many shapes for cursors, even if it is helpful to have some based on magnification. The question came out of me realizing that the cursor precision was going to greatly vary depending on then amount of zoom and I wasn't sure how that played into the initial issue report.

In that case, I think the key info here is the proportions of the cursor elements in relation to one another. Here's how I made it, though I expect the numbers to change since I was not designing these in Qt (in case anyone was unsure 🙃 ). 

I designed this to be 24x24 pixels. The outer lines are 8 pixels long and 2 pixels wide, while the center open box is an 6x6 pixel rectangle. The white outline is only 1 pixel wide on all sides.

<img width="857" alt="Screen Shot 2021-10-08 at 11 06 35 AM" src="https://user-images.githubusercontent.com/50221806/136603146-9f8bcf9c-98fa-4842-8f66-6375c02b40b1.png">

Broken down, that would be:

A 6x6 pixel rectangle with a 2x2 pixel cut out.

<img width="842" alt="Screen Shot 2021-10-08 at 11 08 32 AM" src="https://user-images.githubusercontent.com/50221806/136603176-118b02ba-6bcd-41fd-9096-06f9b1119102.png">

Four 8x2 pixel rectangles centered on each side.

<img width="857" alt="Screen Shot 2021-10-08 at 11 09 04 AM" src="https://user-images.githubusercontent.com/50221806/136603229-d688f64c-5e74-4df1-b414-60f2e9e99f36.png">

A 1 pixel outline on all sides.

<img width="857" alt="Screen Shot 2021-10-08 at 11 09 51 AM" src="https://user-images.githubusercontent.com/50221806/136603257-4c35f720-2a89-45db-b94d-adf676e806f4.png">

I've rounded many areas to match the rounding elsewhere in napari, but this is something that I'm more flexible on. Because it is such a small cursor, it may still blend with the rest of the UI whether or not the rounding is easy to do.

Please let me know whether or not that breakdown is helpful for implementation or if you have other questions or feedback!

**Comment by sofroniewn** [comment]
Those diagrams look great - that should be enough detail for implementation. Maybe that's something @ppwadhwa can take on when she has a chance. I've commented above https://github.com/napari/napari/issues/2142#issuecomment-935467021 on where in the code the current custom cursors are added. We'll probably need to add a new one according to the new designs and then make sure that it is turned on in the correct time. Thanks!!

**Comment by VolkerH** [comment]
Just saw this. 
One comment: if the centre part of the cross hairs cursor is 2x2 or any even number of pixels, it is not clear to me on which pixel that it covers it acts on when the magnification is 1 to 1 (100%). The centre is between those 2x2 pixels. 
I reckon that for this reason the centre part should have an odd number of pixels, either 1x1 or 3x3. 

Other than that I find it very nice.

Edit: I also just noticed that the design shows nicely beveled edges, but when going to pixels they would presumable be square?

**Comment by goanpeca** [comment]
Love this thread and discussion! Thanks everyone. Just wondering for a future round with magnification.

Is this something that should be by default or activated with a modifier like... Cmd/Ctrl to activate magnification, or something that should be configurable with a setting (always magnify vs only with modifier vs never)?

:)

**Comment by isabela-pf** [comment]
> One comment: if the centre part of the cross hairs cursor is 2x2 or any even number of pixels, it is not clear to me on which pixel that it covers it acts on when the magnification is 1 to 1 (100%). The centre is between those 2x2 pixels.

@VolkerH I'm glad you found this issue! 

You are correct! I was designing for visibility in the mockups (my original scale of the cursor with a 1x1 pixel center was way too small over the napari interface), not pixel-perfect accuracy since pixels aren't the most consistent measurement and I do not have information on what the current cursor pixel sizes are in napari. I included the pixels I used just to help give a sense of scale, but I agree that it doesn't make sense to actually have it be 2x2 pixels. Sorry for the confusion.

> I reckon that for this reason the centre part should have an odd number of pixels, either 1x1 or 3x3.

I think 3x3 will be the option that is easier for users to actually see.

This comes back to my question of how this gets generated. My guess is that Qt bases the pixel size on the screen resolution, but I don't know how this plays with the resolution that an image in napari will display at. If the image is very high resolution, it seems like the cursor would still not be precise enough regardless of what size the central opening is (this is why I think the loupe option will be more helpful to users when it is feasible to implement). I may be fundamentally misunderstanding how Qt or napari (or both!) work, however.

> Other than that I find it very nice.

Thanks!

> Edit: I also just noticed that the design shows nicely beveled edges, but when going to pixels they would presumable be square?

Possibly. The beveled edges are a suggestion to match the curves elsewhere in the interface, but they aren't critical to solving the original issue and may not be reasonable to implement.

---

> Just wondering for a future round with magnification.
> Is this something that should be by default or activated with a modifier like... Cmd/Ctrl to activate magnification, or something that should be configurable with a setting (always magnify vs only with modifier vs never)?

@goanpeca I'm going to answer your question like you meant `magnification` as the loupe cursor. Let me know if I misunderstood.

In most interfaces I've encountered (mostly design software for me) that use this type of cursor that magnifies as you move, it is something that appears instantly when the user initiates a selection style, tool, or command that would benefit from it. Like in @emilmelnikov's example above, selecting a color picker tool changes the cursor instantly because it is always helpful to know exactly which pixel is being selected. The only way to remove it is to change to a selection style, tool, or command that does not use that cursor type.

The easy answer would be to have napari follow that pattern. The question is if we have actions in napari that 
1. rely exclusively on a pixel perfect selection 
2. wouldn't be impeded by having a cursor state that takes up a larger portion of the screen. 

I'm in no place to answer these since I'm still pretty new to napari, so any users lurking on this issue can take this as an invite to give their thoughts ❤️.

> activated with a modifier like... Cmd/Ctrl to activate magnification, or something that should be configurable with a setting (always magnify vs only with modifier vs never)?

A modifier sounds like a smart choice! It has its own obstacles to overcome:
- We'd need to figure out how to communicate that modifier to people, and it would probably need to be a persistent communication.
- We'd need to decide if that's always a state the cursor can enter, or if it is just possible when certain selection styles, tools, or commands are selected.
